### PR TITLE
Handle TRUNCATE TABLE on chunks

### LINKED ISSUE #4362: Truncating a compressed chunk does not work, but does not return an error
### What type of bug is this?

Incorrect result

### What subsystems and features are affected?

Compression

### What happened?

Executing TRUNCATE on a compressed chunk returns `TRUNCATE TABLE` but table does not get truncated.

### TimescaleDB version affected

2.5.1 

### PostgreSQL version used

13.5

### What operating system did you use?

Ubuntu 

### What installation method did you use?

Docker

### What platform did you run on?

Microsoft Azure Cloud

### Relevant log output and stack trace

```bash
cscore=# TRUNCATE TABLE   _timescaledb_internal._hyper_314_130697_chunk;
TRUNCATE TABLE
Time: 211.563 ms
cscore=# SELECT COUNT(*) FROM _timescaledb_internal._hyper_314_130697_chunk;
 count 
-------
     1
(1 row)

Time: 177.186 ms
```

### How can we reproduce the bug?

```bash
DROP TABLE IF EXISTS myt;
CREATE TABLE myt (ts timestamptz PRIMARY KEY  , id INTEGER);
SELECT  create_hypertable('myt','ts');
ALTER TABLE myt SET (
  timescaledb.compress
);
INSERT INTO myt VALUES (now(),1);
SELECT compress_chunk(i, if_not_compressed => true) 
  FROM show_chunks('myt') i;
TRUNCATE TABLE   _timescaledb_internal._hyper_314_130697_chunk;
/* We get a response TRUNCATE TABLE */
SELECT COUNT(*) FROM _timescaledb_internal._hyper_314_130697_chunk;
/* Expected 0 , got 1 */
SELECT decompress_chunk(i) FROM show_chunks('myt') i;
TRUNCATE TABLE    _timescaledb_internal._hyper_314_130697_chunk;
SELECT COUNT(*) FROM _timescaledb_internal._hyper_314_130697_chunk;
/* Now we get 0, as expected */
```

### PR DESCRIPTION
**Pull request by mfundul**
Make truncating a uncompressed chunk drop the data for the case where they reside in a corresponding compressed chunk.
    
Generate invalidations for Continuous Aggregates after TRUNCATE, so as to have consistent refresh operations on the materialization hypertable.
    
Fixes #4362

### CODE REVIEW

**Review by svenklemm — CHANGES_REQUESTED** [review]
Why are we blocking this instead of making this work?

**Comment by mfundul** [comment]
> Why are we blocking this instead of making this work?

Various reasons:
1. What's the value of truncating a chunk instead of dropping it? Why keep an empty chunk around?
2. We cannot truncate distributed chunks because they are foreign tables so that is not that consistent of an experience:
```
postgres=# TRUNCATE TABLE _timescaledb_internal._dist_hyper_8_10_chunk;
ERROR:  "_dist_hyper_8_10_chunk" is not a table
```
3. Truncating a chunk does NOT generate continuous aggregate invalidation records today so it is in a fundamentally broken state
4. There are no tests whatsoever for truncating chunks currently so this probably slipped.

Too much work would need to be done to do this properly, for little benefit. Handle CAGG invalidations, local/distributed chunks, compressed/decompressed etc. I don't think it is worth the effort.

**Comment by fabriziomello** [comment]
I agree with @mfundul considerations and even worse will be upgrade current hypertables to also consider the TRUNCATE event on hypertables.

Nowadays the invalidation logs are created by a trigger fired by INSERT/UPDATE/DELETE events on chunks. And to also fire this invalidation trigger by TRUNCATE we'll need to re-create (drop/create) the invalidation trigger in all existing hypertables and chunks and this operation requires AccessExclusiveLock so it can lead to a bad user experience during the extension upgrade (aka locks).

IMHO now the best we can do is block TRUNCATE until we discuss more and figure out a good way to implement it if we agree on how to implement it.

**Review by mkindahl — COMMENTED** [review]
Please fix the following:

- I am lacking a good commit message. It's good to understand that it fixes the issue, but having a description of what this commit does and why will help when working with the commits, especially when you have a lot of commits.
- I am missing a test where you drop a chunk and then try to insert into it again. If there is such a test already, could you add that information to the commit message.

File: src/process_utility.c [hunk 1]
```diff
@@ -1011,6 +1011,15 @@ process_truncate(ProcessUtilityArgs *args)
 		}
 		else
 		{
+			Chunk *chunk = ts_chunk_get_by_relid(relid, false);
+			if (chunk != NULL)
+			{
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("Truncating chunks not supported"),
```
Comment (erimatnor) [inline src/process_utility.c:1019]:
```suggestion
						 errmsg("truncating chunks not supported"),
```

File: src/process_utility.c [hunk 2]
```diff
@@ -1011,6 +1011,15 @@ process_truncate(ProcessUtilityArgs *args)
...
+						 errhint("Please drop the corresponding chunk instead.")));
```
Comment (erimatnor) [inline src/process_utility.c:1020]: No need to be courteous :)

```suggestion
						 errhint("Drop the corresponding chunk instead.")));
```

**Review by erimatnor — COMMENTED** [review]
Reading the discussion, I think we should try to make this work properly, but it is fine to block it for now to ensure correctness. I do lack a proper description/commit message though. A good example of simple change that requires explanation as evident by the discussion.

**Comment by mfundul** [comment]
> I am missing a test where you drop a chunk and then try to insert into it again. If there is such a test already, could you add that information to the commit message.

This commit is about blocking TRUNCATE TABLE on chunks. I am not sure what dropping a chunk would do here, can you elaborate?

File: src/process_utility.c [hunk 3]
```diff
@@ -1011,6 +1011,15 @@ process_truncate(ProcessUtilityArgs *args)
 		}
 		else
 		{
+			Chunk *chunk = ts_chunk_get_by_relid(relid, false);
+			if (chunk != NULL)
+			{
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("truncating chunks not supported"),
```
Comment (erimatnor) [inline src/process_utility.c:1019]: Actually, if I may nitpick: should add "is" to be similar to standard PostgreSQL messages about things not supported:

```suggestion
						 errmsg("truncating chunks is not supported"),
```

**Review by erimatnor — APPROVED** [review]

File: src/process_utility.c [hunk 4]
```diff
@@ -1114,6 +1114,46 @@ process_truncate(ProcessUtilityArgs *args)
 							 */
 							list_changed = true;
 					}
+					else if ((chunk = ts_chunk_get_by_relid(relid, false)) != NULL)
+					{
```
Comment (mkindahl) [inline src/process_utility.c:1118]: Could need a comment here what this is about. This is about truncating a single chunk?

File: src/process_utility.c [hunk 5]
```diff
@@ -1114,6 +1114,46 @@ process_truncate(ProcessUtilityArgs *args)
...
+						/* If the hypertable has continuous aggregates, then invalidate
+						 * the truncated region. */
+						if (ts_continuous_agg_hypertable_status(ht->fd.id) == HypertableIsRawTable)
+						{
+							int64 start = ts_chunk_primary_dimension_start(chunk);
+							int64 end = ts_chunk_primary_dimension_end(chunk);
+
+							Assert(hyperspace_get_open_dimension(ht->space, 0)->fd.id ==
+								   chunk->cube->slices[0]->fd.dimension_id);
+							ts_cm_functions->continuous_agg_invalidate_raw_ht(ht, start, end);
```
Comment (mkindahl) [inline src/process_utility.c:1134]: This is done elsewhere and would probably benefit from having a single function created for this job. Something like `ts_cagg_invalidate_chunk(chunk)`.

File: src/process_utility.c [hunk 6]
```diff
@@ -1114,6 +1114,46 @@ process_truncate(ProcessUtilityArgs *args)
...
+												  NameStr(compressed_chunk->fd.table_name),
+												  -1);
+								MemoryContextSwitchTo(oldctx);
+								list_changed = true;
+							}
+						}
+						list_append = true;
+					}
+					else
+						list_append = true;
```
Comment (mkindahl) [inline src/process_utility.c:1156]: This is horribly hard to follow given the use of `list_append`, `list_changed`, and `rv`. Would benefit from a refactor, but I would accept if you do not think this is in scope for this pull request.

**Review by mkindahl — APPROVED** [review]
The comment to the commit is hard to follow what it actually does. I would be good if you could reformulate it into the problem and how it is solved.

File: src/process_utility.c (hunk 6 above)
Comment (mfundul) [reply src/process_utility.c:1156]: Oh yes, agreed. I found it very difficult to understand how this code actually worked.

**Review by fabriziomello — APPROVED** [review]

**Review by svenklemm — APPROVED** [review]
